### PR TITLE
Wi-Fi: Merge dev branch

### PR DESCRIPTION
**Pull request by krish2718**
This is in preparation for NCS 2.1.0 release, as DEV has diverged the commits are high, but in future where we move to feature branches from DEV branch as rest of Nordic, future PRs are smaller.

### CODE REVIEW

**Comment by krish2718** [comment]
Once https://github.com/nrfconnect/sdk-hostap/pull/14 this is merged, CI should pass.

File: drivers/wifi/nrf700x/Kconfig.nrf700x [hunk 1]
```diff
@@ -14,6 +14,10 @@ config WIFI_NRF700X
 	help
 	  Nordic Wi-Fi Driver
 
+config NRF_WIFI_LOW_POWER
+	bool "Enable low power mode in nRF Wi-Fi chipsets"
+	default n
```
Comment (wbober) [inline drivers/wifi/nrf700x/Kconfig.nrf700x:19]: Any reason not to set this to `y` by default?

Comment (krish2718) [reply drivers/wifi/nrf700x/Kconfig.nrf700x:19]: Just couple of days back got approval from QA that this is working, so, have a local commit which enabled this by default, will update this branch too.

File: drivers/wifi/nrf700x/zephyr/src/qspi/src/rpu_hw_if.c [hunk 2]
```diff
@@ -311,11 +302,13 @@ int rpu_enable(void)
 	rpu_gpio_config();
 	rpu_pwron();
 	rpu_qspi_init();
+	/* WAR for EK boards to wait longer for RPU to be ready after
+	 * powering it on.
+	 */
+	k_msleep(8);
```
Comment (wbober) [inline drivers/wifi/nrf700x/zephyr/src/qspi/src/rpu_hw_if.c:308]: Should this be `SLEEP_TIME_MS` from `rpu_hw_if.h`?

Comment (krish2718) [reply drivers/wifi/nrf700x/zephyr/src/qspi/src/rpu_hw_if.c:308]: Yes, in fact its a slightly different WAR that I have raised in https://github.com/nrfconnect/sdk-nrf/pull/8453 will use that.

**Review by wbober — APPROVED** [review]

File: drivers/wifi/nrf700x/osal/hw_if/hal/src/hal_interrupt.c [hunk 3]
```diff
@@ -456,8 +456,11 @@ static unsigned int hal_rpu_event_get_all(struct wifi_nrf_hal_dev_ctx *hal_dev_c
 			goto out;
 		}
 
-		/* No more events to read */
-		if (!event_addr) {
+		/* No more events to read. Sometimes when low power mode is enabled
+		 * we see a wrong address, but it work after a while, so, add a
+		 * check for that.
+		 */
+		if (!event_addr || event_addr == 0xAAAAAAAA) {
```
Comment (sachinthegreen) [inline drivers/wifi/nrf700x/osal/hw_if/hal/src/hal_interrupt.c:463]: Can 0xAAAAAAAA be replaced with a define?

File: drivers/wifi/nrf700x/zephyr/src/zephyr_disp_scan.c [hunk 4]
```diff
@@ -82,11 +91,16 @@ void wifi_nrf_event_proc_disp_scan_res_zep(void *vif_ctx,
 {
 	struct wifi_nrf_vif_ctx_zep *vif_ctx_zep = NULL;
 	struct umac_display_results *r = NULL;
-	struct wifi_scan_result res;
+	struct wifi_driver_scan_result res;
 	unsigned int i = 0;
+	driver_scan_result_cb_t cb = NULL;
+
+	cb = (driver_scan_result_cb_t) vif_ctx_zep->disp_scan_cb;
```
Comment (sachinthegreen) [inline drivers/wifi/nrf700x/zephyr/src/zephyr_disp_scan.c:98]: Why is this needed?

File: drivers/wifi/nrf700x/zephyr/src/zephyr_disp_scan.c [hunk 5]
```diff
@@ -114,7 +127,7 @@ void wifi_nrf_event_proc_disp_scan_res_zep(void *vif_ctx,
 			res.rssi = (r->signal.signal.unspec_signal);
 		}
 
-		vif_ctx_zep->disp_scan_cb(vif_ctx_zep->zep_net_if_ctx,
+		cb(vif_ctx_zep->zep_net_if_ctx,
 					  0,
 					  &res);
```
Comment (sachinthegreen) [inline drivers/wifi/nrf700x/zephyr/src/zephyr_disp_scan.c:133]: Alignment..

File: samples/wifi/shell/src/wifi_shell.c [hunk 6]
```diff
@@ -234,8 +272,23 @@ static int cmd_supplicant(const struct shell *shell,
...
+			__func__);
+		return -1;
+	}
+
+	ret =  cli_main(argc, argv);
+	if (!ret) {
+		set_dummy_socket_ready();
+	}
+
+	return 0;
```
Comment (sachinthegreen) [inline samples/wifi/shell/src/wifi_shell.c:291]: Should this be doing `return ret`?

**Review by sachinthegreen — CHANGES_REQUESTED** [review]

File: drivers/wifi/nrf700x/osal/hw_if/hal/src/hal_interrupt.c (hunk 3 above)
Comment (krish2718) [reply drivers/wifi/nrf700x/osal/hw_if/hal/src/hal_interrupt.c:463]: Its only used here, and we have magic numbers all over, so, probably take it up in a separate PR?

Comment (sachinthegreen) [reply drivers/wifi/nrf700x/osal/hw_if/hal/src/hal_interrupt.c:463]: Yeah..sounds ok.

**Review by sachinthegreen — APPROVED** [review]

**Review by krga2022 — APPROVED** [review]
Ran following smoke tests and all are passing

1. Scan - PASS
2. 5G_11ax Association/PING in Open -PASS
3. 5G_11ax Association/PING in WPA2-PSK -PASS
4. 5G_11ax Association/PING in WPA3-SAE -PASS
5. 24G_11ax Association/PING in Open -PASS
6. 24G_11ax Association/PING in WPA2-PSK -PASS
7. 24G_11ax Association/PING in WPA3-SAE -PASS

File: drivers/wifi/nrf700x/osal/os_if/inc/osal_api.h [hunk 7]
```diff
@@ -321,6 +313,49 @@ void wifi_nrf_osal_spinlock_irq_rel(struct wifi_nrf_osal_priv *opriv,
...
+ * wifi_nrf_osal_log_dbg() - Log a debug message.
+ * @opriv: Pointer to the OSAL context returned by the @wifi_nrf_osal_init API.
+ * @fmt: Format string.
+ * @...: Variable arguments.
+ *
+ * Logs a debug message.
+ *
+ * Return: Number of characters of the message logged.
+ */
+int wifi_nrf_osal_log_dbg(struct wifi_nrf_osal_priv *opriv,
```
Comment (Damian-Nordic) [inline drivers/wifi/nrf700x/osal/os_if/inc/osal_api.h:327]: Could we keep the previous implementation? This one will make it impossible to compile-out logs based on the configured level and significantly increase the flash usage, right?

Comment (krish2718) [reply drivers/wifi/nrf700x/osal/os_if/inc/osal_api.h:327]: This doesn't change the core logic, just uses the OS abstraction layers instead of directly calling Zephyr's logging APIs.

Comment (Damian-Nordic) [reply drivers/wifi/nrf700x/osal/os_if/inc/osal_api.h:327]: Yes, that's what I'm saying - this abstraction doesn't allow the compiler to compile-out the logs so I wonder if this is really needed.

Comment (krish2718) [reply drivers/wifi/nrf700x/osal/os_if/inc/osal_api.h:327]: Yes, that reverting is not a solution, we need to figure out how to get compiler optimisations with keeping the abstractions.

Comment (Damian-Nordic) [reply drivers/wifi/nrf700x/osal/os_if/inc/osal_api.h:327]: Btw, why is the abstraction needed?

Comment (markaj-nordic) [reply drivers/wifi/nrf700x/osal/os_if/inc/osal_api.h:327]: For 2.1.0 IMHO we should either keep the current implementation which reduces abstraction and flash utilization or revert to the abstract approach but find another way to compile out unnecessary logs. Long story short, we should keep the flash occupancy reduced for the 2.1.0 release. @krish2718 is this what is planned?

File: drivers/wifi/nrf700x/zephyr/src/qspi/src/nrf700x_shell.c [hunk 8]
```diff
@@ -284,9 +284,30 @@ static int cmd_rpuclks_on(const struct shell *shell, size_t argc, char **argv)
 	return shell_ret(rpu_clks_on());
 }
 
+
+void dummy_irq_handler(const struct device *dev, struct gpio_callback *cb, uint32_t pins)
```
Comment (rlubos) [inline drivers/wifi/nrf700x/zephyr/src/qspi/src/nrf700x_shell.c:288]: This should be `static`.

**Review by rlubos — COMMENTED** [review]
@Damian-Nordic @krish2718 There still seem to be some fixes needed on the Matter side? See the twister results.

File: drivers/wifi/nrf700x/osal/os_if/inc/osal_api.h (hunk 7 above)
Comment (krish2718) [reply drivers/wifi/nrf700x/osal/os_if/inc/osal_api.h:327]: > Btw, why is the abstraction needed?

`osal` folder is the agnostic folder common across OSes, BUSes, RPU variants etc.

> Long story short, we should keep the flash occupancy reduced for the 2.1.0 release. @krish2718 is this what is planned?

No, the plan is to fix the `osal` logging flash occupancy post-2.1.0, but if this is blocking then I can revert this commit.

**Comment by krish2718** [comment]
> @Damian-Nordic @krish2718 There still seem to be some fixes needed on the Matter side? See the twister results.

I am working on this locally using matter `lock` sample.

File: drivers/wifi/nrf700x/osal/os_if/inc/osal_api.h (hunk 7 above)
Comment (Damian-Nordic) [reply drivers/wifi/nrf700x/osal/os_if/inc/osal_api.h:327]: IMHO the abstraction should allow OSes inject their declarations, not just definitions. I'm not sure if this is blocking, would need to check what is the size of Matter samples after this change, but I think it would be nice if our samples left some spare flash for customers to experiment with their applications, so every kB is precious.

Comment (krish2718) [reply drivers/wifi/nrf700x/osal/os_if/inc/osal_api.h:327]: Well, that is exactly what is missing in the abstractions, I am fixing the matter build and then I can publish `lock` sample numbers.

**Comment by krish2718** [comment]
Merged all manifest report, updated revisions, this is now good to go.

File: drivers/wifi/nrf700x/osal/os_if/inc/osal_api.h (hunk 7 above)
Comment (krish2718) [reply drivers/wifi/nrf700x/osal/os_if/inc/osal_api.h:327]: IMHo, we do have space, so, we can leave this as is and work on it later?

```
Memory region         Used Size  Region Size  %age Used
           FLASH:      839864 B      1016 KB     80.73%
            SRAM:      378494 B       448 KB     82.51%
        IDT_LIST:          0 GB         2 KB      0.00%
```

Comment (Damian-Nordic) [reply drivers/wifi/nrf700x/osal/os_if/inc/osal_api.h:327]: Thanks for sharing the report. Yeah, I think we should be fine with the numbers for now.

Comment (krish2718) [reply drivers/wifi/nrf700x/osal/os_if/inc/osal_api.h:327]: Reverting that change, slightly increased: 

```
Memory region         Used Size  Region Size  %age Used
           FLASH:      840560 B      1016 KB     80.79%
            SRAM:      378494 B       448 KB     82.51%
        IDT_LIST:          0 GB         2 KB      0.00%

```

**Review by Damian-Nordic — APPROVED** [review]

File: drivers/wifi/nrf700x/osal/os_if/inc/osal_api.h (hunk 7 above)
Comment (krish2718) [reply drivers/wifi/nrf700x/osal/os_if/inc/osal_api.h:327]: @Damian-Nordic We should close this comment and investigate further. Do you agree?

Comment (markaj-nordic) [reply drivers/wifi/nrf700x/osal/os_if/inc/osal_api.h:327]: Ok, these numbers looks good.

Comment (Damian-Nordic) [reply drivers/wifi/nrf700x/osal/os_if/inc/osal_api.h:327]: Yes, I approved the PR already :)

**Review by rlubos — APPROVED** [review]
